### DISCUSSION TITLE
About providing specification for extension.

### DISCUSSION THREAD

**Discussion by ltdrdata** [Comfy-Org/ComfyUI-Manager#3] [Ideas]
This is a simple sketch for specification.
(An extension in the form of git deploy)

1. the specification should be provided as comment in \_\_init\_\_.py 

2. the specification should contains these contents.

* title: title of the extension
* category: basic category of the extension
* tags: for search
* description: brief explanation about the extension
* author: author
* reference: A link where users can find more detailed information about this extension
* node list: A list of nodes provided by this extension
* description of each node: brief explanation of each node
* dependent resources: If there are any resources such as models or other files that need to be downloaded besides the package, they should be listed
* version: version
* dependent extensions
* extension identifier: too recognize each extension
* feature spec list
  * if extension contains multiple selectable features.
  * install method, file for each method

3. file structure

* mandatory: \_\_init\_\_.py
* optional: requirements.txt
* optional: install.py

4. if extension contains .js file.
* install.py should be responsible for copying its own .js file to web/extensions

**Comment by pythongosssss** [comment]
A little bit of feedback from me:

1\. I think putting the extension specification into a json file would be clearer and less error prone than an comment in a py file. 
3\. I dont think \_\_init\_\_.py then needs to be mandatory for js only extensions, and following on...
4\. I dont see why it cant automatically copy js files from a folder specified in the extension spec

You could base the structure off something like npm package.json as I feel its achieving a very similar thing

**Reply by ltdrdata** [reply]
1\. I agree with you. 

3\. If we omit ```__init__.py```, We cannot use the method of using ```git clone``` for custom_nodes. For such extension, We will need another way to check for updates.

4\. Automatically copying .js files was designed as a measure to ensure integrity at runtime after an update, when supporting update functionality based on git. In my opinion, it is desirable to have a simple installation method even if ComfyUI-Manager is not available.

**Reply by pythongosssss** [reply]
OK yeah, I can agree, just adding a simple boilerplate that copies files automatically for both manual clones and the manager is a good idea.

For something like my custom scripts repo, I feel like people would want to pick and choose specific ones as they do widely different things, would you consider adding some kind of "feature selection" when installing an extension? Or the alternative is for me to split my repo into multiple repos (which is also fine)

**Reply by pythongosssss** [reply]
I guess having optional installation would need to be implemented by the extension itself as I cant really see a generic way to do that that would work for both manual and automatic installs. I'll probably split my repo up

**Reply by ltdrdata** [reply]
> OK yeah, I can agree, just adding a simple boilerplate that copies files automatically for both manual clones and the manager is a good idea.
> 
> For something like my custom scripts repo, I feel like people would want to pick and choose specific ones as they do widely different things, would you consider adding some kind of "feature selection" when installing an extension? Or the alternative is for me to split my repo into multiple repos (which is also fine)

It would be better to clone the repo in a managed space other than custom_nodes, and implement it in a form that allows users to select features based on the .json spec in the repo and install them. It looks like a new view extension is needed.
